File: starter-bots/KotlinBot/src/main/kotlin/arctica/service/BotService.kt
package arctica.service

import BotDto
import arctica.service.model.CommandAction
import arctica.service.model.PlayerCommand
import arctica.service.model.ScoutTower
import arctica.service.model.dto.GameStateDto
import arctica.service.model.dto.Map
import arctica.service.model.dto.ResourceNode
import arctica.service.model.enums.ActionType
import arctica.service.model.enums.ResourceType
import java.util.*
import kotlin.random.Random

class BotService {

    private lateinit var gameState: GameStateDto
    private lateinit var botId: UUID
    private lateinit var bot: BotDto
    private lateinit var worldMap: Map
    private lateinit var scoutedTowers: ScoutTower

    private fun updateBotState(gameState: GameStateDto) {
        this.gameState = gameState
        this.botId = gameState.botId
        this.bot = gameState.bots.first { it.id == botId }
        this.worldMap = gameState.world.map
    }

    fun computeNextPlayerAction(): PlayerCommand? {

        println("AVAILABLE UNITS ${bot.availableUnits}")
        if (bot.availableUnits < 1) return null

        val command = if (bot.map.nodes.isEmpty()) {

            scoutedTowers = worldMap.scoutTowers.random()

            listOf(CommandAction(ActionType.SCOUT.value, 1, scoutedTowers.id))

        } else {

            val targetNode = bot.map.nodes.random()
            val node: ResourceNode = gameState.world.map.nodes.find { it.id == targetNode } ?: return null

            val action = when (node.type) {
                ResourceType.WOOD.value -> ActionType.LUMBER
                ResourceType.FOOD.value -> ActionType.FARM
                ResourceType.STONE.value -> ActionType.MINE
                else -> {
                    return null
                }
            }

            println("Sending: ${action.name}")

            listOf(
                CommandAction(
                    action.value,
                    Random.nextInt(
                        bot.availableUnits
                    ),
                    node.id
                )
            )
        }
        return PlayerCommand(botId, command)
    }

    companion object {
        fun updateBotService(botService: BotService?, gameState: GameStateDto) {
            botService?.updateBotState(gameState)
        }
    }
}

File: starter-bots/NETCoreBot/NETCoreBot/Services/BotService.cs
using Domain.Models;
using NETCoreBot.Enums;
using NETCoreBot.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace NETCoreBot.Services
{
    public class BotService
    {
        private BotDto _bot;
        private CommandAction _playerAction;
        private PlayerCommand _playerCommand;
        private GameState _gameState;
        private EngineConfigDto _engineConfigDto;

        public BotService()
        {
            _playerAction = new CommandAction();
            _playerCommand = new PlayerCommand();
            _gameState = new GameState();
            _bot = new BotDto();
            _engineConfigDto = new EngineConfigDto();
        }

        public BotDto GetBot()
        {
            return _bot;
        }

        public PlayerCommand GetPlayerCommand()
        {

            PlayerCommand playerCommand = new PlayerCommand();

            if (this._gameState.World != null)
            {
                if (this._gameState.World.Map.Nodes.Count > 0)
                {
                    playerCommand.Actions.Add(new CommandAction()
                    {
                        Type = ActionType.Farm,
                        Units = this._gameState.Bots.FirstOrDefault(go => go.Id == _bot.Id).AvailableUnits,
                        Id = this._gameState.World.Map.Nodes[0].Id,
                    });

                    playerCommand.PlayerId = this._bot.Id;
                }
                else
                {
                    playerCommand.Actions.Add(new CommandAction()
                    {
                        Type = ActionType.Scout,
                        Units = 1,
                        Id = this._gameState.World.Map.ScoutTowers[0].Id,
                    });

                    playerCommand.PlayerId = this._bot.Id;
                }
            }

            return playerCommand;
        }

        public void SetBot(BotDto bot)
        {
            _bot.Id = bot.Id;
        }

        public void ComputeNextPlayerAction(PlayerCommand playerCommand)
        {

            _playerAction = playerCommand.Actions[0];
        }

        public GameState GetGameState()
        {
            return _gameState;
        }

        public void SetGameState(GameState gameState)
        {
            _gameState = gameState;
            UpdateSelfState();
        }

        private void UpdateSelfState()
        {
            _bot = _gameState.Bots.FirstOrDefault(go => go.Id == _bot.Id);
        }      
        
        public EngineConfigDto GetEngineConfigDto()
        {
            return _engineConfigDto;
        }

        public void SetEngineConfigDto(EngineConfigDto engineConfigDto)
        {
            _engineConfigDto = engineConfigDto;
        }

    }
}